src/Lexer.ts

```diff
@@ -0,0 +1,98 @@
+// @author Anush kumar N <anush.kumar@codingmart.com>
+
+import TokenType from './Token';
+
+type arrayOfString = Array<string>;
+type literalWithType = { type: Symbol; literal: string | undefined };
+type identifyLiteralReturnType = { token: literalWithType; updatedPos: number | undefined };
+
+// Performs lexical analysis on raw string input.
+function* Lexer(input: string) {
+  const inputAsArr: arrayOfString = Array.from(input);
+  inputAsArr.push(TokenType.eof.toString());
+
+  let position: number = 0;
+  for (; position < inputAsArr.length; position++) {
+    const str: string = inputAsArr[position];
+    const { token, updatedPos } = identifyLiteral({ str, position, inputAsArr });
+    updatedPos && (position = updatedPos);
+
+    yield token;
+  }
+}
+
+// identifyLiteral helps to identify literals i.e either the token, number or identifier
+type identifyLiteralParams = {
+  str: string;
+  position: number;
+  inputAsArr: arrayOfString;
+};
+const identifyLiteral = (data: identifyLiteralParams): identifyLiteralReturnType => {
+  const { str, position, inputAsArr } = data;
+  let token, updatedPos;
+
+  switch (str) {
+    case '.':
+      token = { type: TokenType.dot, literal: str };
+      break;
+    case '[':
+      token = { type: TokenType.lbracket, literal: str };
+      break;
+    case ']':
+      token = { type: TokenType.rbracket, literal: str };
+      break;
+    default:
+      if (isNumeric(str)) {
+        const { updatedPosInt, int } = readNumber(position, inputAsArr) || {};
+        updatedPosInt && (updatedPos = updatedPosInt);
+        token = { type: TokenType.int, literal: int };
+      } else {
+        const { updatedPosIdent, ident } = readIdentifier(position, inputAsArr) || {};
+        updatedPosIdent && (updatedPos = updatedPosIdent);
+        token = { type: TokenType.ident, literal: ident };
+      }
+  }
+
+  return { token, updatedPos };
+};
+
+// isNumeric helps identify if the given string is a number or not
+const isNumeric = (n: string): boolean => {
+  const num: number = parseFloat(n);
+
+  return !isNaN(num) && isFinite(num);
+};
+
+// readNumber reads the integer next to the token
+type readNumberReturnType = { int: string; updatedPosInt: number } | undefined;
+const readNumber = (pos: number, srcInput: arrayOfString): readNumberReturnType => {
+  let int: string = '';
+  const prevToken: string = srcInput[pos - 1];
+  for (; pos < srcInput.length; pos++) {
+    if (prevToken === '[' && srcInput[pos] === ']') {
+      return { int, updatedPosInt: pos - 1 };
+    }
+    int += srcInput[pos];
+  }
+};
+
+// readIdentifier reads the identfier next to the token
+type readIdentifierReturnType = { ident: string; updatedPosIdent: number } | undefined;
```

Comment:
The reference spec guarantees the return value will be a token. Can't we do the same here?

Reply:
yes sure